File: 06-Abstract-Factory/src/cpp/main.cpp
// Chapter 06 — Abstract Factory Pattern in C++
// Compile: g++ -std=c++17 -o main main.cpp && ./main

#include <iostream>
#include <memory>
#include "UIFactory.h"

// Client function — works with UIFactory interface only
void buildUI(const UIFactory& factory) {
    auto button = factory.createButton();
    auto checkbox = factory.createCheckbox();
    auto menu = factory.createMenu();

    std::cout << "--- Rendering UI ---\n";
    button->render();
    checkbox->render();
    menu->render();

    std::cout << "--- Interacting ---\n";
    button->onClick();
    checkbox->toggle();
    menu->select("File > Save");
}

int main() {
    std::cout << "=== Windows UI ===\n";
    WindowsFactory windowsFactory;
    buildUI(windowsFactory);

    std::cout << "\n=== Mac UI ===\n";
    MacFactory macFactory;
    buildUI(macFactory);

    // Runtime family selection
    std::cout << "\n=== Runtime Selection (simulated) ===\n";
    std::string os = "mac";
    std::unique_ptr<UIFactory> factory;
    if (os == "windows") {
        factory = std::make_unique<WindowsFactory>();
    } else {
        factory = std::make_unique<MacFactory>();
    }
    buildUI(*factory);

    return 0;
}

File: 06-Abstract-Factory/src/rust/main.rs
// Chapter 06 — Abstract Factory Pattern in Rust
// Run: rustc main.rs && ./main

mod products;
mod factory;

use factory::*;

// Client function — works with UIFactory trait only
fn build_ui(factory: &dyn UIFactory) {
    let button = factory.create_button();
    let checkbox = factory.create_checkbox();
    let menu = factory.create_menu();

    println!("--- Rendering UI ---");
    button.render();
    checkbox.render();
    menu.render();

    println!("--- Interacting ---");
    button.on_click();
    checkbox.toggle();
    menu.select("File > Save");
}

fn main() {
    println!("=== Windows UI ===");
    build_ui(&WindowsFactory);

    println!("\n=== Mac UI ===");
    build_ui(&MacFactory);

    // Runtime family selection
    println!("\n=== Runtime Selection (simulated) ===");
    let os = "mac";
    let factory: Box<dyn UIFactory> = if os == "windows" {
        Box::new(WindowsFactory)
    } else {
        Box::new(MacFactory)
    };
    build_ui(&*factory);
}
